compiler/resolution/caches.cpp

```diff
@@ -94,3 +99,422 @@ static bool isCacheEntryMatch(SymbolMap* s1, SymbolMap* s2) {
+// and add it to all cacheInfos in 'parentInfos'.
+static void updateOneCandidate(std::vector<GenericsCacheInfo*>& parentInfos,
+                               FnSymbol* candidFn, VisibilityInfo& visInfo) {
+  if (candidFn->hasFlag(FLAG_COBEGIN_OR_COFORALL) ||
+      candidFn->hasFlag(FLAG_BEGIN) ||
+      candidFn->hasFlag(FLAG_ON))
+    return; // do not record
+
+  CalledFunInfo cfi(candidFn->name, getVisibilityScope(candidFn->defPoint),
+                    visInfo.visitedScopes);
+
+  // Performance opt: skip this loop if the previous "best" candidate
+  // had the same name and scope. ("scope" means its defPoint->parentExpr.)
+
+  for_vector(GenericsCacheInfo, pInfo, parentInfos)
+    pInfo->infos.push_back(cfi);
+}
+
+//
+// The call 'visInfo.call' resolves to the candidate(s) visible through
+// (one of) its POIs. So, 'visInfo.call' must be in an instantiation
+// of a generic function; we have cached this instantiation in genericsCache.
+// For this cache entry to be applicable to another call in the future, those
+// candidates will need to be visible from that call or its POI(s).
+//
+// Record this requirement in the cacheInfos of the FnSymbols containing
+// visInfo.call and the POI(s). The requirement represented as a CalledFunInfo
+// within FnSymbol->cacheInfo. The CalledFunInfo is generated based on
+// 'visInfo' in updateOneCandidate().
+//
+void updateCacheInfosForACall(VisibilityInfo& visInfo,
+                              ResolutionCandidate* best1,
+                              ResolutionCandidate* best2,
+                              ResolutionCandidate* best3) {
+  INT_ASSERT(visInfo.poiDepth <= (int)visInfo.instnPoints.size());
+
+  //
+  // 'parentInfos' contains the cacheInfos to be updated.
+  // We precompute this set as an optimization, so that updateOneCandidate
+  // can compute the CalledFunInfo once and reuse it for all cacheInfos.
+  //
+  std::vector<GenericsCacheInfo*> parentInfos;
+  parentInfos.reserve(visInfo.poiDepth);
+
+  if (!addOneParentInfo(parentInfos, visInfo.call))
+    // We hit a method-like function. Currently these are always applicable.
+    // Do not impose CFI constraint on it or its callers.
+    return;
+
+  for (int depth = 0; depth < visInfo.poiDepth - 1; depth++)
+    if (!addOneParentInfo(parentInfos, visInfo.instnPoints[depth]))
+      break;
+
+  if (best1) updateOneCandidate(parentInfos, best1->fn, visInfo);
+  if (best2) updateOneCandidate(parentInfos, best2->fn, visInfo);
+  if (best3) updateOneCandidate(parentInfos, best3->fn, visInfo);
+}
+
+//
+// isApplicableInstantiation() and helpers
+//
+
+//
+// Check how visInfo.visitedScopes, starting from 'startVS', cover the CFIs
+// in 'toProcess'. Nullify the CFIs that have been satisfied and decrement
+// 'remainingCFIs' correspondingly. If we encounter a scope that might make
+// this cached instantiation not applicable, return false.
+//
+static bool reportVisitedScopes(std::vector<CalledFunInfo*>& toProcess,
```

Comment:
Maybe this should be called `recordVisitedScopes` ? or `eliminateVisitedScopes` ?